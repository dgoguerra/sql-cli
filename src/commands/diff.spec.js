const { getTestKnex, runCli } = require("../../test/utils");
const { hydrateKnex } = require("../knexUtils");

const TEST_DATETIME_1 = "2020-07-24 18:34:00";
const TEST_DATETIME_2 = "2020-07-24 19:25:00";

const TEST_TABLE1_CONTENT = [
  { field_1: 12, field_2: "foo", created_at: TEST_DATETIME_1 },
  { field_1: 30, field_2: "bar", created_at: TEST_DATETIME_1 },
];

const TEST_TABLE2_CONTENT = [
  { field_1: 12.3, field_2: "foo", created_at: TEST_DATETIME_1 },
  { field_1: 30.45, field_2: "bar", created_at: TEST_DATETIME_2 },
];

describe("diff", () => {
  let knex1;
  let connUri1;

  let knex2;
  let connUri2;

  beforeAll(async () => {
    knex1 = hydrateKnex(getTestKnex("diff-1.db"));
    connUri1 = knex1.getUri();

    knex2 = hydrateKnex(getTestKnex("diff-2.db"));
    connUri2 = knex2.getUri();

    await knex1.schema.createTable("table_1", (t) => {
      t.increments("id");
      t.integer("field_1");
      t.text("field_2");
      t.timestamps();
      t.index(["field_1"]);
    });
    await knex2.schema.createTable("table_2", (t) => {
      t.increments("id");
      t.decimal("field_1").notNullable();
      t.text("field_2").defaultTo("default text");
      t.timestamps();
      t.unique(["field_1"]);
    });
    await knex2.schema.createTable("table_3", (t) => {
      t.bigIncrements("id_field");
      t.bigInteger("field_1");
      t.timestamps();
    });

    await knex1("table_1").insert(TEST_TABLE1_CONTENT);
    await knex2("table_2").insert(TEST_TABLE2_CONTENT);
  });

  it("can diff tables with changes", async () => {
    const output = await runCli(`diff ${connUri1}/table_1 ${connUri2}/table_2`);
    expect(output).toMatchSnapshot();
  });

  it("can diff tables with no changes", async () => {
    const output = await runCli(`diff ${connUri1}/table_1 ${connUri1}/table_1`);
    expect(output).toMatchSnapshot();
  });

  it("can diff tables with no changes, showing all", async () => {
    const output = await runCli(
      `diff ${connUri1}/table_1 ${connUri1}/table_1 --all`
    );
    expect(output).toMatchSnapshot();
  });

  it("can diff tables data with changes", async () => {
    const output = await runCli(
      `diff ${connUri1}/table_1 ${connUri2}/table_2 --data`
    );
    expect(output).toMatchSnapshot();
  });

  it("can diff tables data with no changes", async () => {
    const output = await runCli(
      `diff ${connUri1}/table_1 ${connUri1}/table_1 --data`
    );
    expect(output).toMatchSnapshot();
  });

  it("can diff tables data with no changes, showing all", async () => {
    const output = await runCli(
      `diff ${connUri1}/table_1 ${connUri1}/table_1 --data --all`
    );
    expect(output).toMatchSnapshot();
  });

  it("can diff schemas with changes", async () => {
    const output = await runCli(`diff ${connUri1} ${connUri2}`);
    expect(output).toMatchSnapshot();
  });

  it("can diff schemas with no changes", async () => {
    const output = await runCli(`diff ${connUri1} ${connUri1}`);
    expect(output).toMatchSnapshot();
  });

  it("can diff schemas with no changes, showing all", async () => {
    const output = await runCli(`diff ${connUri1} ${connUri1} --all`);
    expect(output).toMatchSnapshot();
  });
});
