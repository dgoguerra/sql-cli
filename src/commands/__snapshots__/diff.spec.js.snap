// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff can diff schemas 1`] = `
"table   rows bytes  columns    indexes
table_1 2    4.1 kB 5x deleted 1x deleted
table_2 2    4.1 kB 5x created 1x created
table_3 0    4.1 kB 4x created none

Tables: 1x deleted, 2x created
"
`;

exports[`diff can diff schemas with no changes 1`] = `
"Tables: 1x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`diff can diff schemas with no changes, showing all 1`] = `
"table   rows bytes  columns    indexes
table_1 2    4.1 kB 5x similar 1x similar

Tables: 1x similar
"
`;

exports[`diff can diff tables 1`] = `
"column  type            nullable     default               foreign key
field_1 integer → float true → false
field_2 text            true         [null] → default text [null] → table_1.field_1

index                  algorithm unique columns
table_1_field_1_index  unknown   false  field_1
table_2_field_1_unique unknown   true   field_1

Columns: 2x changed, 3x similar (hidden)
Indexes: 1x deleted, 1x created
Re-run with --all to show hidden rows
"
`;

exports[`diff can diff tables data 1`] = `
"Diff of tables content (rows 1 to 20):

id field_1    created_at
1  12 → 12.3  2020-07-24 18:34:00
2  30 → 30.45 2020-07-24 18:34:00 → 2020-07-24 19:25:00

Rows: 2x changed
"
`;

exports[`diff can diff tables data with all options 1`] = `
"Diff of tables content (rows 2 to 2):

id field_1
2  30 → 30.45

Rows: 1x changed
"
`;

exports[`diff can diff tables data with no changes 1`] = `
"Diff of tables content (rows 1 to 20):

Rows: 2x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`diff can diff tables data with no changes, showing all 1`] = `
"Diff of tables content (rows 1 to 20):

id
1
2

Rows: 2x similar
"
`;

exports[`diff can diff tables with no changes 1`] = `
"Columns: 5x similar (hidden)
Indexes: 1x similar (hidden)
Re-run with --all to show hidden rows
"
`;

exports[`diff can diff tables with no changes, showing all 1`] = `
"column     type     nullable default foreign key
id         integer  false
field_1    integer  true
field_2    text     true
created_at datetime true
updated_at datetime true

index                 algorithm unique columns
table_1_field_1_index unknown   false  field_1

Columns: 5x similar
Indexes: 1x similar
"
`;
